### DISCUSSION TITLE
Where's the request scoping?

### DISCUSSION THREAD

**Discussion by scampos1030** [pallets-eco/flask-sqlalchemy#1077] [Q&A] [answered]
I've been told all up and down the internet over the years that one of the beauties of flask-sqlalchemy over the generic version is that it handles niceties like automatically scoping sessions to requests (even better than scoping to threads).

But from reading the code all I can find is an attempt to scope to threads.
https://github.com/pallets-eco/flask-sqlalchemy/blob/main/src/flask_sqlalchemy/__init__.py#L36-L39

I'm sure I missed something but does the code somewhere swap in a more specific scope_func, or are we just working on the assumption that thread scope will get us the same effect.

**Comment by ThiefMaster** [comment]
You don't have multiple requests in parallel in the same thread so yes, I think you assumption is correct.

**Reply by scampos1030** [reply]
Maybe.  Couple things to consider though.

1. There are some (although contrived) scenarios where different threads might want to operate on the same request/session and thread-scoping would prevent them from doing so.  Sure the session isn't designed for concurrency, but if using locks (maybe some remaining DB operations get offloaded from the main thread?) it might be sequential operation.
2. This library doesn't use pure threads for scoping.  It uses greenlets.  There are some pre-fork scenarios where different child processes reference the same non-existent, implicit thread ([link](https://github.com/python-greenlet/greenlet)).  The result would be multiple processes operating on the same session despite being responsible for different requests.  I think even guinocorn supports a pre-fork model.  Code example below:

```
from multiprocessing import process
from greenlet import getcurrent
import time
import os
from multiprocessing import Process

def printit():
    print(getcurrent())
    time.sleep(3)
    print(os.getpid())
    time.sleep(3)

tasks = [Process(target=printit) for i in range(4)]

for t in tasks:
    t.start()

for i in tasks:
    t.join()
```
Output:
```
<greenlet.greenlet object at 0x7efc468f5930 (otid=0x7efc468d3e40) current active started main>
<greenlet.greenlet object at 0x7efc468f5930 (otid=0x7efc468d3e40) current active started main>
<greenlet.greenlet object at 0x7efc468f5930 (otid=0x7efc468d3e40) current active started main>
<greenlet.greenlet object at 0x7efc468f5930 (otid=0x7efc468d3e40) current active started main>
706896
706899
706898
706897
```

**Comment by davidism** [comment, marked as answer]
The session is scoped to the Flask application context in Flask-SQLAlchemy 3. Because an active context is required, the session will always be cleaned up after it is used. The context is unique to each thread/greenlet/coroutine because it uses `contextvars`.
